bench/graphcc_bench.cpp

```diff
@@ -0,0 +1,180 @@
+#include <benchmark/benchmark.h>
+#include <xxhash.h>
+#include <iostream>
+#include <unistd.h>
+#include <fstream>
+
+#include "binary_graph_stream.h"
+#include "bucket.h"
+#include "test/sketch_constructors.h"
+
+constexpr uint64_t KB   = 1024;
+constexpr uint64_t MB   = KB * KB;
+constexpr uint64_t seed = 374639;
+
+// Linux-only, flush the filesystem cache
+// requires sudo privileges :(
+static void flush_filesystem_cache() {
+  sync();
+  std::ofstream drop("/proc/sys/vm/drop_caches");
+  if (drop.is_open()) {
+    drop << "3" << std::endl;
+  }
+  else {
+    std::cout << "WARNING: could not drop filesystem cache. BM_FileIngest will be inaccurate. ";
+    std::cout << "Running as root may be required." << std::endl;
+  }
+}
+
+// Test the speed of reading all the data in the kron16 graph stream
+static void BM_FileIngest(benchmark::State &state) {
+  // determine the number of edges in the graph
+  uint64_t num_edges;
+  {
+    BinaryGraphStream stream("/mnt/ssd2/binary_streams/kron_16_stream_binary", 1024);
+    num_edges = stream.edges();
+  }
+
+  // flush fs cache
+  flush_filesystem_cache();
+  
+  // perform benchmark
+  for (auto _ : state) {
+    BinaryGraphStream stream("/mnt/ssd2/binary_streams/kron_16_stream_binary", state.range(0));
+
+    uint64_t m = stream.edges();
+    GraphUpdate upd;
+    while (m--) {
+      benchmark::DoNotOptimize(upd = stream.get_edge());
+    }
+  }
+  state.counters["Ingestion_Rate"] = benchmark::Counter(state.iterations() * num_edges, benchmark::Counter::kIsRate);
+}
+BENCHMARK(BM_FileIngest)->RangeMultiplier(2)->Range(KB << 2, MB / 4);
+
+// Test the speed of hashing using a method that loops over seeds and a method that 
+// batches by seed
+// The argument to this benchmark is the number of hashes to batch
+static void BM_Hash_XXH64(benchmark::State &state) { 
+  uint64_t num_seeds = 8;
+  uint64_t num_hashes = state.range(0);
+  uint64_t output;
+  for (auto _ : state) {
+    for (uint64_t h = 0; h < num_seeds; h++) {
+      for (uint64_t i = 0; i < num_hashes; i++) {
+        benchmark::DoNotOptimize(output = XXH64(&i, sizeof(uint64_t), seed + h));
```

Comment:
why no optimization?

Reply:
So benchmark::DoNotOptimize() prevents the results of the expression from being optimized out. Optimizations still happen within the expression ie XXH64 still gets optimized but the result is forced to be written to output.

Reply:
That prevents the compiler from going "Oh you never actually use 'output'. Let me just optimize out all of this work and make the benchmark a no-op"